### PR TITLE
HummingBird version1.0

### PR DESCRIPTION
**Pull request by paying4529-zz**
Complete basic function

### CODE REVIEW

File: extension.json [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  "apiVersion": "1.0.0",
+  "author": "Peggy Lin",
+  "version": "1.0.0",
+  "description": "I2C SPEC testing",
+  "extensions": {
+    "HummingBird": {
+      "type": "AnalogMeasurement",
+      "entryPoint": "hummingbird.HummingBird",
+      "metrics": {
+      	"spec": {
```
Comment (p9n) [inline extension.json]: fix the indent

File: extension.json [hunk 2]
```diff
@@ -0,0 +1,160 @@
+{
+  "name": "HummingBird",
```
Comment (p9n) [inline extension.json:2]: I guess we need a more descriptive name if we want to publish.

Should be okay right now.

File: generate_report.py [hunk 3]
```diff
@@ -0,0 +1,532 @@
+"""HummingBird Report methods.
+
+The file contain methods related to output report.
+
+"""
+import datetime
+import tempfile
+temp_path = tempfile.gettempdir()
+
+__main__ = ["SVGfile", "OutputReportFile"]
```
Comment (p9n) [inline generate_report.py]: what's __main__ for?

File: generate_report.py [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+    rect_x = rect_idx//rate*5 - rect_width
+    svgfile += (
+        f"<rect x={rect_x} y='0' width={rect_width} height='100%' fill='red' opacity='0.3'/>"
+    )
+
+  # Data Polyline
+
+  points = ""
+  for i in range(0, len(data), rate):
+    points += f"{i//rate*5},{(data_max - int(data[i]*50))*2 + 40} "
```
Comment (p9n) [inline generate_report.py]: I guess the string concat is slow when data is very large. You can try to output to the html file directly

File: generate_report.py [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+def SVGfile(data, data_max, data_min, rect_idx, rect_width, field):
+  """Generate SVG plot.
+
+  Args:
+    data:  numpy array of voltages
+    data_max:  data maximum value
+    data_min:  data minimum value
+    rect_idx:  index where the worst pattern occurs
+    rect_width:  the index width where the worst pattern last
+    field:  the SPEC field of this SVF plot
```
Comment (p9n) [inline generate_report.py]: SVF -> SVG

File: generate_report.py [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+  }
+}
+</script>"""
+  report += style + "</style></head><body>" + script
+
+  if mode is None:
+    mode = "Unknown ( Note: SCL data has not been specified... )"
+
+  time_now = datetime.datetime.now()
+  report_path = temp_path+f"/{time_now.strftime('%Y-%m-%d-%H%M%S')}report.html"
```
Comment (p9n) [inline generate_report.py]: use `os.path.join(temp_path, f"...")`

File: hummingbird.py [hunk 7]
```diff
@@ -0,0 +1,979 @@
...
+
+The project serves as an extension measurement module
+run on Saleae Logic2 Software. Supporting main function
+would be running I2C electrical test on capture data.
+
+"""
+import datetime
+import os
+import tempfile
+#raise Exception(os.path.dirname(__file__))
```
Comment (p9n) [inline hummingbird.py]: remove this line

File: hummingbird.py [hunk 8]
```diff
@@ -0,0 +1,979 @@
...
+        dt = datetime.datetime.strptime(dt, "%Y-%m-%d %H:%M:%S")
+        ms, us = int(subms[0:3]), int(subms[3:6])
+        ns, ps = int(subms[6:9]), int(subms[9:12])
+        self.scl_start_time = GraphTime(dt, millisecond=ms, microsecond=us,
+                                        nanosecond=ns, picosecond=ps)
+        self.scl_sampling_period = float(head[1].split("\n")[0])
+        self.f_clk = float(head[2].split("\n")[0])
+        self.scl_data = np.loadtxt(self.scl_data_path, delimiter=",",
+                                   skiprows=3)
+        f.close()
```
Comment (p9n) [inline hummingbird.py]: no need to close since you are using `with`.

File: hummingbird.py [hunk 9]
```diff
@@ -0,0 +1,979 @@
...
+    """
+    self.samples.append(data.samples)
+    if self.sampling_period is None:
+      self.sampling_period = (
+          ((data.end_time - data.start_time) / data.sample_count).__float__())
+
+    if self.start_time is None:
+      self.start_time = data.start_time
+
+  def measure(self):
```
Comment (p9n) [inline hummingbird.py]: Split this into small functions, i.e. write a function for each step

File: hummingbird.py [hunk 10]
```diff
@@ -0,0 +1,979 @@
...
+          svg_fields[f+"_sda"] = SVGfile(self.sda_data[start_idx:end_idx],
+                                         sda_v_max, sda_v_min, idx-start_idx,
+                                         svgwidth[f], f+"_sda")
+
+      ############### Generate and Show Report ##############
+
+      report_path = OutputReportFile(
+          self.mode, spec_limit.copy(), vs, values.copy(), result.copy(),
+          fail.copy(), num_pass, svg_fields, uni_addr)
+      os.system(f"open {report_path}")
```
Comment (p9n) [inline hummingbird.py]: Use [subprocess.run(["open", report_path])](https://docs.python.org/3/library/subprocess.html#subprocess.run)

File: hummingbird.py [hunk 11]
```diff
@@ -0,0 +1,979 @@
...
+                  ("_percent" not in param) and ("_idx" not in param))}
+      passes = {param: result for (param, result) in result.items()
+                if ((result == 0) and ("_margin" not in param) and
+                    ("_percent" not in param) and ("_idx" not in param))}
+      num_pass = len(passes)
+      values["spec"] = len(fail)
+
+    uni_addr = list(set(addr_list))
+    uni_addr = [f"0x{int(addr, 2):02X}" for addr in uni_addr]
+    uni_addr = list(set(uni_addr))
```
Comment (p9n) [inline hummingbird.py]: no need because duplicated addresses in `uni_addr` is already removed in line 907

**Review by p9n — CHANGES_REQUESTED** [review]
Just putting some general comments here, still need some time to review the measure() function.

- Variable naming:
    - Use upper case for global constants like `TEMP_PATH`
    - The "f" in `SVGfile` should be upper case
- https://google.github.io/styleguide/pyguide.html#36-whitespace
    -  Surround binary operators with a single space on either side
- Google now recommends 4 space indent. I know this is really hard to change at this time, so this is optional :p
- You can put the parameter type in function signature, for example `def SVGfile(data: np.array, data_max: float, data_min: float, ...)`
- (bonus item) try to upload a generated html to [w3c validator](https://validator.w3.org/) to check if it has any syntax error

**Comment by paying4529-zz** [comment]
- indent, parameter type, html format, variables, uni_addr, subprocess fixed
- change type decision flow
- save output report to HummingBird local downloaded directory

**Comment by paying4529-zz** [comment]
- split measure into small functions

**Comment by paying4529-zz** [comment]
- directly write report rather than concat string

File: generate_report.py [hunk 12]
```diff
@@ -0,0 +1,547 @@
...
+          f"<div id='{field}_hide' class='hide'>"
+          "<div class='column_left'>SCL capture</div><div class='column_right'>"
+      )
+    svgfile += (
+        f"<svg viewBox='0 0 {width} {height * 1.2}' height={height * 0.7} width=auto>"
+    )
+
+    # Red Rect to Mark the Measure Area
+
+    rect_width = max(rect_width//rate * 5, 7)
```
Comment (p9n) [inline generate_report.py]: whitespace around `//`, also at line 73 and 82

**Review by p9n — COMMENTED** [review]
Looks really good now, thanks.
One more thing: remove the .pyc files.

Also there's some future tasks, no need to do in this PR:
- Add __pycache__ to .gitignore to prevent uploading unwanted files again
- Add unittests

File: README.md [hunk 13]
```diff
@@ -1 +1,38 @@
-# Automated Electrical Test for I2C Signal
+# HummingBird Automated Electrical Test for I2C Signal
+  
+An extension that check whether the captured analog measurements meet I2C SPEC limitation.
+
+SPEC reference:	[NXP UM10204](https://www.nxp.com/docs/en/user-guide/UM10204.pdf)
+
+A report html file would be generated at /var/tmp/report.html
```
Comment (p9n) [inline README.md:7]: this line needs update

**Comment by paying4529-zz** [comment]
- delete pyc file
- add whitespace

**Review by p9n — APPROVED** [review]

**Pull request merged**